File: 프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations

def solution(k, dungeons):
    answer = 0
    sunseo = [i for i in range(0,len(dungeons))]
    sunseo = list(permutations(sunseo, len(sunseo)))

    for current_sunseo in sunseo :
        tmp = k
        cnt = 0
        
        for idx in current_sunseo :
            if dungeons[idx][0] > tmp :
                break
                
            tmp -= dungeons[idx][1]
            cnt += 1
            
        answer = max(answer, cnt)
        
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    int answer = -1;
    boolean[] visit;
    
    public int solution(int k, int[][] dungeons) {
        int n = dungeons.length;
        visit = new boolean[n];
        
        dfs(k, dungeons, 0, n);

        return answer;
    }
    
    public void dfs(int k, int[][] dungeons, int depth, int n) {
        answer = Math.max(answer, depth);
        for (int i=0; i<n; i++) {
            if (k < dungeons[i][0] || visit[i]) {
                continue;
            }
            visit[i] = true;
            dfs(k - dungeons[i][1], dungeons, depth+1, n);
            visit[i] = false;
        }
        return;
    }
}